RHDEVDOCS-3423-managing-user-defined-alerting-rules (#48754)
openshift/openshift-docs: modules/monitoring-modifying-core-platform-alerting-rules.adoc

// Module included in the following assemblies:
//
// * monitoring/managing-alerts.adoc

:_content-type: PROCEDURE
[id="modifying-core-platform-alerting-rules_{context}"]
= Modifying core platform alerting rules

As a cluster administrator, you can modify core platform alerting rules to suit your needs. For example, you can tweak an alert expression or change the severity label of an alert.

.Prerequisites

* You have access to the cluster as a user with the `cluster-admin` role.

> * You have installed the OpenShift CLI (`oc`).

Reviewer: I think we should add
edit featuregate to enable TechPreview feature
`# oc edit featuregate/cluster`
then change the spec to:
```
spec:
  featureSet: TechPreviewNoUpgrade
```

Author: Added a step to enable TP features and added a new line to the sample per your feedback.

.Procedure

. Create a new YAML configuration file named `example-modified-alerting-rule.yaml` in the `openshift-monitoring` namespace.

. Add an `AlertRelabelConfig` resource to the YAML file. 
The following example modifies the `severity` setting to `critical` for the default platform `watchdog` alerting rule: 
+
[source,yaml]
----
apiVersion: monitoring.openshift.io/v1alpha1
kind: AlertRelabelConfig
metadata:
  name: watchdog
  namespace: openshift-monitoring
spec:
  configs:
  - sourceLabels: [alertname,severity] <1>
    regex: "Watchdog;none" <2>
    targetLabel: severity <3>
    replacement: critical <4>
    action: Replace <5>
----
<1> The source labels for the values to be modified.

> <2> The name of the default alerting rule you want to modify.

Reviewer: change to
`The name of the alerting rule you want to modify.`

Author: Changed

<3> The target label of the value you want to modify.

> <4> The new label to replace the target label.

Reviewer: The new value to replace the target label.

Author: changed

> <5> The relabel action that replaces the old value with the the new value. This `action` name must be `Replace`.

Reviewer: This `action` name must be `Replace`.
change to
The default `action` is `Replace`.
```
# oc explain AlertRelabelConfig.spec.configs.action
KIND:     AlertRelabelConfig
VERSION:  monitoring.openshift.io/v1alpha1

FIELD:    action <string>

DESCRIPTION:
     action to perform based on regex matching. Must be one of: replace, keep,
     drop, hashmod, labelmap, labeldrop, or labelkeep. Default is: 'replace'
```

Author: Fixed and added a sentence listing the other possible values.

// Module included in the following assemblies:
//
// * monitoring/managing-alerts.adoc

:_content-type: PROCEDURE
[id="modifying-core-platform-alerting-rules_{context}"]
= Modifying core platform alerting rules

> As a cluster administrator, you can modify core platform alerting rules to suit your needs. For example, you can tweak an alert expression or change the severity label of an alert.

Reviewer 2: ```suggestion
As a cluster administrator, you can modify core platform alerts before they are routed by Alertmanager to suit your needs. For example, you can change the severity label of an alert, add a custom label or exclude and alert from being send to Alertmanager.
```
The AlertRelabelConfig can not change the alert expressions as this is done after an alert already fires. Hope these changes clarify that.

Author: Revised these sentences using your suggestion as the basis: 

```
As a cluster administrator, you can modify core platform alerts before Alertmanager routes them to a receiver. 
For example, you can change the severity label of an alert, add a custom label, or exclude an alert from being sent to Alertmanager.
```

* You have access to the cluster as a user with the `cluster-admin` role.
* You have installed the OpenShift CLI (`oc`).
* You have enabled Technology Preview features, and all nodes in the cluster are ready.

.Procedure

. Create a new YAML configuration file named `example-modified-alerting-rule.yaml` in the `openshift-monitoring` namespace.

. Add an `AlertRelabelConfig` resource to the YAML file. 
The following example modifies the `severity` setting to `critical` for the default platform `watchdog` alerting rule: 
+
[source,yaml]
----
apiVersion: monitoring.openshift.io/v1alpha1
kind: AlertRelabelConfig
metadata:
  name: watchdog
  namespace: openshift-monitoring
spec:
  configs:
  - sourceLabels: [alertname,severity] <1>
    regex: "Watchdog;none" <2>
    targetLabel: severity <3>
    replacement: critical <4>
    action: Replace <5>
----
<1> The source labels for the values to be modified.
<2> The name of the alerting rule you want to modify.
<3> The target label of the value you want to modify.
<4> The new value to replace the target label.
<5> The relabel action that replaces the old value with the the new value based on regex matching.
The default action is `Replace`.

> Other possible values are `Keep`, `Drop`, `Hashmod`, `Labelmap`, `Labeldrop`, and `Labelkeep`.

Reviewer: with the fix from https://bugzilla.redhat.com/show_bug.cgi?id=2100312#c9
Other possible values are `Keep`, `Drop`, `HashMod`, `LabelMap`, `LabelDrop`, and `LabelKeep`.

Author: fixed

> Other possible values are `Keep`, `Drop`, `HashMod`, `LabelMap`, `LabelDrop`, and `LabelKeep`.

Reviewer 2: ```suggestion
Other possible values are `Keep`, `Drop`, `HashMod`, `LabelMap`, `LabelDrop`, and `LabelKeep`.
The `Drop` action will prevent a firing alert to be delivered to Alertmanager and as a consequence will stop any notification routing for this alert.
```
A little clarification based on the addition above.

Author: Added:

`The `Drop` action prevents a firing alert from being delivered to Alertmanager and, as a result, stops any notification routing for this alert.
`

// Module included in the following assemblies:
//
// * monitoring/managing-alerts.adoc

:_content-type: PROCEDURE
[id="modifying-core-platform-alerting-rules_{context}"]
= Modifying core platform alerting rules

As a cluster administrator, you can modify core platform alerting rules to suit your needs. For example, you can tweak an alert expression or change the severity label of an alert.

.Prerequisites

* You have access to the cluster as a user with the `cluster-admin` role.
* You have installed the OpenShift CLI (`oc`).
* You have enabled Technology Preview features, and all nodes in the cluster are ready.

.Procedure

. Create a new YAML configuration file named `example-modified-alerting-rule.yaml` in the `openshift-monitoring` namespace.

. Add an `AlertRelabelConfig` resource to the YAML file. 
The following example modifies the `severity` setting to `critical` for the default platform `watchdog` alerting rule: 
+
[source,yaml]
----
apiVersion: monitoring.openshift.io/v1alpha1
kind: AlertRelabelConfig
metadata:
  name: watchdog
  namespace: openshift-monitoring
spec:

>   featureSet: TechPreviewNoUpgrade

Reviewer: remove `featureSet: TechPreviewNoUpgrade`

Author: fixed

  configs:
  - sourceLabels: [alertname,severity] <1>
    regex: "Watchdog;none" <2>
    targetLabel: severity <3>
    replacement: critical <4>
    action: Replace <5>
----
<1> The source labels for the values to be modified.
<2> The name of the alerting rule you want to modify.
<3> The target label of the value you want to modify.
<4> The new value to replace the target label.
<5> The relabel action that replaces the old value with the the new value based on regex matching.
The default action is `Replace`.

> Other possible values are `keep`, `drop`, `hashmod`, `labelmap`, `labeldrop`, and `labelkeep`.

Reviewer: @simonpasquier 
actions from
```
# oc explain AlertRelabelConfig.spec.configs.action
KIND:     AlertRelabelConfig
VERSION:  monitoring.openshift.io/v1alpha1

FIELD:    action <string>

DESCRIPTION:
     action to perform based on regex matching. Must be one of: replace, keep,
     drop, hashmod, labelmap, labeldrop, or labelkeep. Default is: 'replace'
```
and https://github.com/openshift/api/pull/1227 used the capitalized actions, do we need to change to below?
Other possible values are `Keep`, `Drop`, `Hashmod`, `Labelmap`, `Labeldrop`, and `Labelkeep`.

Author: I changed these value names to the capitalized actions published in the OpenShift API.

As a cluster administrator, you can modify core platform alerts before Alertmanager routes them to a receiver. 
For example, you can change the severity label of an alert, add a custom label, or exclude an alert from being sent to Alertmanager.

.Prerequisites

* You have access to the cluster as a user with the `cluster-admin` role.
* You have installed the OpenShift CLI (`oc`).
* You have enabled Technology Preview features, and all nodes in the cluster are ready.

.Procedure

. Create a new YAML configuration file named `example-modified-alerting-rule.yaml` in the `openshift-monitoring` namespace.

. Add an `AlertRelabelConfig` resource to the YAML file. 
The following example modifies the `severity` setting to `critical` for the default platform `watchdog` alerting rule: 
+
[source,yaml]
----
apiVersion: monitoring.openshift.io/v1alpha1
kind: AlertRelabelConfig
metadata:
  name: watchdog
  namespace: openshift-monitoring
spec:
  configs:
  - sourceLabels: [alertname,severity] <1>
    regex: "Watchdog;none" <2>
    targetLabel: severity <3>
    replacement: critical <4>
    action: Replace <5>
----
<1> The source labels for the values to be modified.

> <2> The name of the alerting rule you want to modify.

Reviewer: according to https://github.com/openshift/api/blob/master/monitoring/v1alpha1/0000_50_monitoring_02_alertrelabelconfigs.crd.yaml#L64-L66
we'd better change to:
regex against which the value of source labels is matched.

Reviewer 2: yeah good catch!

Author: Changed to:

`The relabel action that replaces the old value with the new value. The new value is based on the regex against which the value of the source labels is matched.`

<3> The target label of the value you want to modify.
<4> The new value to replace the target label.

> <5> The relabel action that replaces the old value with the the new value based on regex matching.

Reviewer 3: ```suggestion
<5> The relabel action that replaces the old value with the new value based on regex matching.
```

Author: fixed

The default action is `Replace`.
Other possible values are `Keep`, `Drop`, `HashMod`, `LabelMap`, `LabelDrop`, and `LabelKeep`.

> The `Drop` action prevents a firing alert from being delivered to Alertmanager and, as a result, stops any notification routing for this alert.

Reviewer: I think it's no need to explain `Drop` action separately, we also have other possible actions but not explained here

Reviewer 2: I'd leave it up to the professional writers whether this makes sense here or only adds confusion. We mention this behavior in another place: https://github.com/openshift/openshift-docs/pull/48754/files#diff-e1e506f67411279fb9dd1d14e895b35f0b3ab44af9774073cb306c120da16561R27

The motivation was to mention the somewhat strange behavior of `Drop` when it comes to the alert being shown in the UI, but not routed to notification channels.
But I agree its slightly awkward here, that is my bad.

Author: I think @juzhao is correct. It can safely be omitted here since it is mentioned elsewhere and because we're not defining the other options anyway.
